Prioritized URLRequest signature, organized overloads

bdrelling opened the pull request:
**Description**
The intent of this PR is to change the primary `request` and `requestPublisher` signatures for `HTTPClient` to prefer `URLRequest` instead of any custom object for this SDK or a collection of properties to build a request. This allows us to pass in existing `URLRequest` objects without issue. 

In addition, I added a convenience initializer to `URLRequest` and cleaned up a bit of the shared code as well. Hopefully, this makes the primary `HTTPClient` and `HTTPClient+Combine` classes more maintainable overall.

bdrelling commented on Sources/UtilityBeltNetworking/Extensions/DataResponse+Failure.swift:
```diff
@@ -0,0 +1,13 @@
+// Copyright © 2021 SpotHero, Inc. All rights reserved.
+
+extension DataResponse {
```
This extension was copied from `HTTPClient` and moved into its own file.

bdrelling commented on Sources/UtilityBeltNetworking/Extensions/URLRequest+Init.swift:
```diff
@@ -0,0 +1,30 @@
+
+import Foundation
+
+public extension URLRequest {
```
This extension was based on `HTTPClient.configuredURLRequest`.

khaptonstall commented on Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestOverloads.swift:
```diff
@@ -0,0 +1,260 @@
+        do {
+            convertedRequest = try request.asURLRequest()
+        } catch {
+            DispatchQueue.main.async {
```
```suggestion
            dispatchQueue.async {
```

khaptonstall commented on Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:
```diff
@@ -0,0 +1,233 @@
+        /// - Parameter request: An object defining properties to make the request with.
+        /// - Parameter validators: An array of validators that will be applied to the response.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on.
```
```suggestion
        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on. Defaults to `.main`.
```

khaptonstall commented on Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:
```diff
@@ -0,0 +1,233 @@
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
+        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on.
```
```suggestion
        /// - Parameter dispatchQueue: The `DispatchQueue` to make the request on. Defaults to `.main`.
```

khaptonstall commented on Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:
```diff
@@ -0,0 +1,233 @@
+        /// - Parameter request: The `URLRequest` to make the request with.
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
```
This should come after the `dispatchQueue` to match the method signature.

khaptonstall replied:
```suggestion
        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data. Defaults to `JSONDecoder()`.
```

khaptonstall commented on Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:
```diff
@@ -0,0 +1,233 @@
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter dispatchQueue: The dispatch queue on which the response will be published. Defaults to `.main`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
```
```suggestion
        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data. Defaults to `JSONDecoder()`.
```

khaptonstall commented on Sources/UtilityBeltNetworking/Extensions/HTTPClient+RequestPublisherOverloads.swift:
```diff
@@ -0,0 +1,233 @@
+        /// - Parameter validators: An array of validators that will be applied to the response. Defaults to ensuring a JSON mime type.
+        /// - Parameter interceptor: An object that can intercept the url request. Defaults to `nil`.
+        /// - Parameter dispatchQueue: The dispatch queue on which the response will be published. Defaults to `.main`.
+        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data.
```
```suggestion
        /// - Parameter decoder: The `JSONDecoder` to use when decoding the response data. Defaults to `JSONDecoder()`.
```

bdrelling commented:
@khaptonstall Thanks for the thorough review, updated all the outdated comments I could find across those four files.

Regarding the test failures, this was because of the `HTTPClientTests` file, which was primarily handling timeouts. Previously, it would use the `HTTPClient` method of configuring a `URLRequest`, but given that this method was removed in favor of a direct initializer on `URLRequest`, I updated the test cases to work with the `URLRequest` object and a new set of extensions directly.
